RxWebKit/Sources/WKNavigationDelegateEvents+Rx.swift

```diff
@@ -0,0 +1,170 @@
+            .methodInvoked(.didReceiveServerRedirectForProvisionalNavigation)
+            .map(navigationEventWith)
+        return ControlEvent(events: source)
+    }
+    
+    /// Reactive wrapper for delegate method `webView(_ webView: WKWebView, didFail navigation: WKNavigation!, withError error: Error)`.
+    public var didFailNavigation: ControlEvent<WKNavigationFailEvent> {
+        let source: Observable<WKNavigationFailEvent> = delegate
+            .methodInvoked(.didFailNavigation)
+            .map(navigationFailEventWith)
+        return ControlEvent(events: source)
+    }
+    
+    /// Reactive wrapper for delegate method `webView(_ webView: WKWebView, didFailProvisionalNavigation navigation: WKNavigation!, withError error: Error)`.
+    public var didFailProvisionalNavigation: ControlEvent<WKNavigationFailEvent> {
+        let source: Observable<WKNavigationFailEvent> = delegate
+            .methodInvoked(.didFailProvisionalNavigation)
+            .map(navigationFailEventWith)
+        return ControlEvent(events: source)
+    }
+    
+    /// Reactive wrapper for delegate method `webView(_ webView: WKWebView, didReceive challenge: URLAuthenticationChallenge, completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void)`
+    public var didReceiveChallenge: ControlEvent<WKNavigationChallengeEvent> {
+        /// __ChallengeHandler is same as ChallengeHandler
+        /// They are interchangeable, __ChallengeHandler is for internal use.
+        // ChallengeHandler is exposed to the user on subscription.
+        typealias __ChallengeHandler =  @convention(block) (URLSession.AuthChallengeDisposition, URLCredential?) -> Void
+        /*! @abstract Invoked when the web view needs to respond to an authentication challenge.
+         @param webView The web view that received the authentication challenge.
+         @param challenge The authentication challenge.
+         @param completionHandler The completion handler you must invoke to respond to the challenge. The
+         disposition argument is one of the constants of the enumerated type
+         NSURLSessionAuthChallengeDisposition. When disposition is NSURLSessionAuthChallengeUseCredential,
+         the credential argument is the credential to use, or nil to indicate continuing without a
+         credential.
+         @discussion If you do not implement this method, the web view will respond to the authentication challenge with the NSURLSessionAuthChallengeRejectProtectionSpace disposition.
+         */
+        let source: Observable<WKNavigationChallengeEvent> = delegate
+            .sentMessage(.didReceiveChallenge)
+            .map { arg in
+                let view = try castOrThrow(WKWebView.self, arg[0])
+                let challenge = try castOrThrow(URLAuthenticationChallenge.self, arg[1])
+                /**
+                 This is a holy grail part for more information please read the following articles.
```

Comment:
awesome

Reply:
Great work here - I would add some comments into the implementation works since there is no way to understand it by just reading the code. The links you provided might go down, leaving future contributors uncertain of what this does.